lib/client-assets.php

```diff
@@ -208,6 +208,40 @@ function gutenberg_preload_api_request( $memo, $path ) {
 
 	$response = rest_do_request( $request );
 	if ( 200 === $response->status ) {
+
+		// Makes sure to check User's capabilities for the Post type being edited.
+		if ( '/wp/v2/users/me?context=edit' === $path ) {
+			global $post;
+
+			// Get the Post's post type.
+			$post_type = get_post_type( $post );
+
+			// Use the REST API User's Controller caps as default ones.
+			$post_type_caps = (array) $response->data['capabilities'];
+
+			if ( post_type_exists( $post_type ) ) {
+				// The Post Type object contains the Post Type's specific caps.
+				$post_type_object = get_post_type_object( $post_type );
+
+				// Loop in the Post Type's caps to validate the User's caps for it.
+				foreach ( $post_type_object->cap as $post_cap => $post_type_cap ) {
+					// Ignore caps requiring a post ID.
+					if ( in_array( $post_cap, array( 'edit_post', 'read_post', 'delete_post' ) ) ) {
+						continue;
+					}
+
+					// Set or reset User's cap for the post type.
+					$post_type_caps[ $post_cap ] = current_user_can( $post_type_cap );
+				}
+
+				$response->set_data(
+					array(
+						'capabilities' => $post_type_caps,
```

Comment:
Really awesome work here Thanks.

As far as I understand, these capabilities are "post type" specific capabilities only. Do you think it makes more sense to use a more descriptive key `post_type_capabilties => array( $post_type => $post_type_caps )` to avoid confusion on the client-side?

Granted this means we need to update some client code as well.

Also this raises the question of whether this should be a separate endpoint `/users/me/post_type_caps?postType=post`

Reply:
You're very welcome @youknowriad 

I agree it makes a lot of sense. As I lack some knowledge of React, I took no risk adventuring myself into this part of Gutenberg's code :smiley:

As you said, it would require some updates into the client code so that the`post_type_capabilties` are used if available.

About the new endpoint, yes it's a possibility. You can alternatively use a specific query parameter `/users/me/?context=edit&post_type=post` and temporarly use the `rest_prepare_user` filter to include the `post_type_capabilties` into the response before editing the `WP_REST_Users_Controller::prepare_item_for_response()` method once Gutenberg is merged into core.

Reply:
> About the new endpoint, yes it's a possibility. You can alternatively use a specific query parameter /users/me/?context=edit&post_type=post and temporarly use the rest_prepare_user filter to include the post_type_capabilties into the response before editing the WP_REST_Users_Controller::prepare_item_for_response() method once Gutenberg is merged into core.

That sounds like a good solution for now. We'll probably have to bring this issue to the REST API team later. Might be good to have @aduth's insights

Reply:
Also, the client changes should be small enough, I can help with those if necessary ;)